### PR TITLE
Add Computable-reals library

### PR DESCRIPTION
**Pull request by Izaakwltn**
This adds a computable-reals library, wrappers around functions from: https://github.com/stylewarning/computable-reals

### CODE REVIEW

File: library/reals/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+The primary type is a `Creal`, a wrapper on the computable-reals type of the same name.
+
+As stated in the computable-reals README:
+
+"Computable real numbers `x` are interpreted as (potentially) infinite fractions in base 2 that are specified through a rule for computation of an integer `a` with `|(2^k)*x - a| <= 1` for any `k>=0`."
+
+### Instances:
+
+`EQ` and `Ord` have been implemented in order to fulfill the needs of the `Num` typeclass, though they are not exact. Their definitions are shaky because they can only be verified up to an arbitrary, specified precision. The precision threshold for `Creal` comparison is stored in the lisp variable `*creal-comparison-threshold*`, which can be accessed using the library functions `creal-prec` and changed using `set-creal-prec`. The default precision is `106`.
```
Comment (stylewarning) [inline library/reals/README.md:16]: I wouldn't mention the Lisp variable name, and the Coalton function names need changing.

File: library/reals/coalton-reals.lisp [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+;;;
+;;; Defining the Creal type as a wrapper around computable-reals
+;;;
+
+(coalton-toplevel
+  
+  (repr :native cr:creal)
+  (define-type Creal)
+
+  (define (set-comparison-threshold! k)
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:36]: Return Unit.

File: library/reals/coalton-reals.lisp [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    "Sets the global `Creal` comparison threshold to k bits after the 'decimal' point.
+
+See `comparison-threshold` for more details."
+    (lisp UFix (k)
+      (cl:setq *creal-comparison-threshold* k)))
+
+  (define (comparison-threshold) ;; radix point
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:43]: stray comment

File: library/reals/coalton-reals.lisp [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+             LT)))))
+
+(coalton-toplevel
+  
+  (define-instance (Reciprocable Creal)
+    (define (/ a b)
+      (lisp Creal (a b)
+        (cr:/r a b)))
+    (define (math:reciprocal n)
+      (/ 1 n))))
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:99]: use /r with one argument

File: library/reals/coalton-reals.lisp [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+    (define ee (exp 1))))
+
+(coalton-toplevel
+  
+  (define-instance (math:Radical Creal)
+
+    (define (math:nth-root n x)
+      (lisp Creal (n x)
+        (cr:Expt-r x (cl:/ 1 n))))
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:129]: (cl:/ n)

File: library/reals/coalton-reals.lisp [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+(coalton-toplevel
+
+  (define-instance (Complex Creal)
+    (define (Complex a b)
+      (complex::%Complex a b))
+
+    (define (real-part z)
+      (match z
+        ((coalton-library/math/complex::%Complex a _) a)))
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:174]: package name

**Review by stylewarning — CHANGES_REQUESTED** [review]
a few things to start

File: library/reals/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+The primary type is a `Creal`, a wrapper on the computable-reals type of the same name.
+
+As stated in the computable-reals README:
+
+"Computable real numbers `x` are interpreted as (potentially) infinite fractions in base 2 that are specified through a rule for computation of an integer `a` with `|(2^k)*x - a| <= 1` for any `k>=0`."
+
+### Instances:
+
+`EQ` and `Ord` have been implemented in order to fulfill the needs of the `Num` typeclass, though they are not exact. Their definitions are shaky because they can only be verified up to an arbitrary, specified precision. The precision threshold for `Creal` can be accessed using the library function `comparison-threshold` and changed using `set-comparison-threshold!`. The default precision is `106`.
```
Comment (stylewarning) [inline library/reals/README.md:16]: nit: `Eq` not `EQ`

File: library/reals/coalton-reals.lisp [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        x))))
+
+(coalton-toplevel
+  
+  (define-instance (Ord Creal)
+    (define (<=> a b)
+      (let diff-approxd = (lisp Integer (a b)
+                            (cr:APPROX-R (cr:-R a b)
+                                         *creal-comparison-threshold*)))
+      (cond ((math:zero? diff-approxd)
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:86]: nit: newline after cond

File: library/reals/coalton-reals.lisp [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+    (define pi
+      (lisp Creal ()
+        cr:+pi-r+))))
+
+
+(coalton-toplevel
+
+  (define-instance (Complex Creal)
+    (define (Complex a b)
```
Comment (stylewarning) [inline library/reals/coalton-reals.lisp:171]: nit: lowercase `c` for the `complex` method

**Review by stylewarning — APPROVED** [review]
nit: i think the creal functions should be lowercased, but it won't block the PR.

Since I only had nits, I can merge tomorrow, but you can make the tweaks before then if you want.

**Comment by eliaslfox** [comment]
I'll look at this tomorrow.

**Comment by Izaakwltn** [comment]
> nit: i think the creal functions should be lowercased, but it won't block the PR.
> 
> Since I only had nits, I can merge tomorrow, but you can make the tweaks before then if you want.

Nits picked

File: coalton.asd [hunk 10]
```diff
@@ -205,6 +205,18 @@
                (:file "impl-default"
                 :if-feature (:or :coalton-portable-bigfloat (:not :sbcl)))))
 
+(asdf:defsystem #:coalton/library/reals
```
Comment (eliaslfox) [inline coalton.asd:208]: `#:coalton-library/computable-reals`

File: coalton.asd [hunk 11]
```diff
@@ -205,6 +205,18 @@
...
+  :description "A Coalton interface for computable-reals (https://github.com/stylewarning/computable-reals)"
+  :author "Coalton contributors (https://github.com/coalton-lang/coalton)"
+  :license "MIT"
+  :version (:read-file-form "VERSION.txt")
+  :pathname "library/reals"
```
Comment (eliaslfox) [inline coalton.asd:213]: `library/computable-reals`

File: library/reals/lisp-reals.lisp [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+(in-package #:coalton-library/lisp-reals)
+
+(defvar *creal-comparison-threshold* 106)
+
+;;;
+;;; Converting objects into creals
+;;;
+
+(defgeneric into-creal (x)
```
Comment (eliaslfox) [inline library/reals/lisp-reals.lisp:20]: Don't define generic functions to call from Coalton. Just call the implementation directly in the instance definition.

**Review by eliaslfox — APPROVED** [review]

**Pull request merged**